src/udiff/_diff_array.py

```diff
@@ -125,27 +160,34 @@ def _backward_jacobian(self, grad_variables, end_node, position, base):
                     self._jacobian[end_node][position] + grad_variables
                 )
 
-        if self._vjp:
-            diffs = list(self._vjp(grad_variables))
+        if self._vjp_func:
+            diffs = list(self._vjp_func(grad_variables))
             for pn, p in self._parents:
                 p._backward_jacobian(diffs[pn], end_node, position, base)
 
     def to(self, x, grad_variables=None, jacobian=False):
-        if jacobian:
-            if x._jacobian is None or self not in x._jacobian:
-                for position in itertools.product(*[range(i) for i in self.shape]):
-                    grad_variables = np.zeros_like(self.value)
-                    grad_variables._value[position] = 1
-                    self._backward_jacobian(grad_variables, self, position, base=x)
-
-            x.jacobian[self] = np.reshape(
-                np.stack(x.jacobian[self].values()), self.shape + x.shape
-            )
-            return x.jacobian[self]
-        else:
-            if x._diff is None or self not in x._diff:
-                self.backward(grad_variables, base=x)
-            return x._diff[self]
+        if self.mode == "vjp":
+            if jacobian:
+                if x._jacobian is None or self not in x._jacobian:
+                    for position in itertools.product(*[range(i) for i in self.shape]):
+                        grad_variables = np.zeros_like(self.value)
+                        grad_variables.value[position] = 1
+                        self._backward_jacobian(grad_variables, self, position, base=x)
+
+                x._jacobian[self] = np.reshape(
+                    np.stack(x._jacobian[self].values()), self.shape + x.shape
+                )
+                return x._jacobian[self]
+            else:
+                if x._vjp_diff is None or self not in x._vjp_diff:
+                    self.backward(grad_variables, base=x)
+                return x._vjp_diff[self]
+        elif self.mode == "jvp":
+            if jacobian or grad_variables:
+                raise NotImplementedError(
+                    "JVP does not yet support grad_variables, jacobian and higher order derivative"
+                )
```

Comment:
Will you resolve this limitation in this PR or another one?

Reply:
These limitations are caused by JVP calculating the value and diff at the same time in the forward process. I will resolve the `jacobian` and `grad_variables` in another pr. While we may use a compromise method to use `grad_variables` such as `x.set_grad_variables(grad_variables)`. But it will lead to some functions and atrributes used in JVP that will not be used in VJP. So I'm considering to split the DIffArray into two different classes: `VJPDiffArray` and `JVPDiffArray`. I don't have a good idea for the higher-order derivative support now. I plan to solve it in September.